### PR TITLE
Nim 0.19/devel compatibility

### LINKED ISSUE #41: Nim 0.19
I'm just wondering if the bindings are going to be updated to compile with the latest nim version 0.19. As of now, there is the error:

> godotapigen.nim(9, 26) Error: type mismatch: got \<string>
but expected one of:
proc getIdent(ic: IdentCache; identifier: string): PIdent
proc getIdent(ic: IdentCache; identifier: string; h: Hash): PIdent
proc getIdent(ic: IdentCache; identifier: cstring; length: int; h: Hash): PIdent
>
> expression: getIdent(ident)

which can be solved as mentioned in https://github.com/pragmagic/godot-nim/issues/32#issuecomment-435637015. But after that, there are still errors regarding string nil checks (strings can't be nil anymore in 0.19). And even if I use the option --nilseqs:on for backwards compatibility other errors show up. I would love to see the Nim bindings updated,

### PR DESCRIPTION
**Pull request by bluenote10**
Attempt to fix #41

I addressed the most obvious issues regarding Nim 0.19. 

Unfortunately, I still can't compile the nim-godot-stub, because there is now a very strange macro error:

```
fpscounter.nim(7, 7) template/generic instantiation of `gdobj` from here
fpscounter.nim(10, 3) Warning: the '.this' pragma is deprecated [Deprecated]
fpscounter.nim(7, 7) template/generic instantiation of `gdobj` from here
../../godot-nim/godot/nim/godotmacros.nim(513, 23) Error: type mismatch: got <int literal(0)> but expected 'bool'
```

This is referring to the code line:

```nimrod
  result.add(getAst(
    registerGodotClass(classNameIdent, classNameLit, isRef, parentName,
                       genSym(nskProc, "createFunc"), obj.isTool)))
```

As far as I can see from stdout debugging none of the arguments is an `int literal (0)`. Is this a Nim bug?

### CODE REVIEW

**Comment by bluenote10** [comment]
The internal error was caused by this Nim issue: https://github.com/nim-lang/Nim/issues/10136

Fortunately it should be simple to work-around the issue, because the `toGodotName` probably can be a `proc` instead of a `macro`.

Now I'm again stuck with the issue in the first post :(.

**Comment by zetashift** [comment]
Trying to run this with the latest devel which fixed the typedesc `is` issue. Currently getting this error:
```
..\_godotapi\input.nim(88, 34) template/generic instantiation of `getSingleton` from here
..\..\..\..\..\.nimble\pkgs\godot-0.7.18\nim\godotnim.nim(369, 30) Error: type mismatch: got <string>
but expected one of:
macro asCString(s: static[string]): cstring

expression: asCString(toGodotName(T))
       Tip: 11 messages have been suppressed, use --verbose to show them.
     Error: Execution failed with exit code 1
        ... Command: "C:\Users\rskar\.nimble\bin\nim.exe" c --noNimblePath --path:"C:\Users\rskar\.nimble\pkgs\godot-0.7.18" --path:"C:\Users\rskar\.nimble\pkgs\compiler-0.19.9" "-o:..\_dlls\nim_64.dll"  "..\src\stub.nim" 
```
Also got a nakefile.nim error but that was solvable by remove the isNil check.

**Comment by RSDuck** [comment]
I don't know if that's the expected behaviour, but I think the problem is that the return value of toGodotName is set to string. This way the returned string literal is immedietly treated as a non const(aka static) string. It's probably solved by changing the return type to untyped.

**Comment by zetashift** [comment]
the `is issue has been solved in latest devel though: https://github.com/nim-lang/Nim/commit/ab99bdfc408e56a2e8a1a3ada21effe983433851

**Comment by narimiran** [comment]
> compiles now on devel and runs fine.

Any news on this one?

We would like to include `godot-nim` in our test suite to ensure we don't introduce some breaking changes in the future, but for that to happen `godot-nim` first needs to compile on current devel (or 0.19 branch).

**Comment by bluenote10** [comment]
It's been a while that I worked on this, but from my point of view it was ready to merge...

@endragor doesn't seem very active recently. Anyone else with merge rights on this repo?

File: godot/nim/godotmacros.nim [hunk 1]
```diff
@@ -197,7 +197,7 @@ proc parseType(definition, callSite: NimNode): ObjectDecl =
       parseError(option, "valid type specifier expected")
   result.isTool = isTool
 
-  if result.parentName.isNil: result.parentName = "Object"
+  if result.parentName == "": result.parentName = "Object"
```
Comment (endragor) [inline godot/nim/godotmacros.nim]: Use `.len == 0` check to keep this compatible with earlier versions of Nim.

File: godot/nim/godotmacros.nim [hunk 2]
```diff
@@ -447,7 +444,7 @@ proc genType(obj: ObjectDecl): NimNode {.compileTime.} =
   let objTy = newNimNode(nnkObjectTy)
   typeDef.add(newNimNode(nnkRefTy).add(objTy))
   objTy.add(newEmptyNode())
-  if obj.parentName.isNil:
+  if obj.parentName == "":
```
Comment (endragor) [inline godot/nim/godotmacros.nim]: Use `.len == 0` check to keep this compatible with earlier versions of Nim.

File: godot/nim/godotmacros.nim [hunk 3]
```diff
@@ -506,23 +503,25 @@ proc genType(obj: ObjectDecl): NimNode {.compileTime.} =
     result.add(meth.nimNode)
 
   # Register Godot object
-  let parentName = if obj.parentName.isNil: newStrLitNode("Object")
+  let parentName = if obj.parentName == "": newStrLitNode("Object")
```
Comment (endragor) [inline godot/nim/godotmacros.nim]: Use `.len == 0` check.

File: godot/nim/godotmacros.nim [hunk 4]
```diff
@@ -506,23 +503,25 @@ proc genType(obj: ObjectDecl): NimNode {.compileTime.} =
...
                    else: newStrLitNode(obj.parentName)
   let classNameLit = newStrLitNode(obj.name)
   let classNameIdent = ident(obj.name)
-  let isRef: bool = if obj.parentName.isNil: false
+  let isRef: bool = if obj.parentName == "": false
```
Comment (endragor) [inline godot/nim/godotmacros.nim]: Use `.len == 0` check.

File: godot/nim/godotnim.nim [hunk 5]
```diff
@@ -204,7 +202,7 @@ macro baseNativeType(T: typedesc): cstring =
     if typeName == "NimGodotObject":
       break
     t = getType(t[1][1])
-  if baseT.isNil:
+  if baseT == "":
```
Comment (endragor) [inline godot/nim/godotnim.nim]: Use `.len == 0` check.

File: godot/nim/godotnim.nim [hunk 6]
```diff
@@ -345,26 +343,20 @@ proc newRStrLit(s: string): NimNode {.compileTime.} =
   result = newNimNode(nnkRStrLit)
   result.strVal = s
 
-static:
-  import strutils
-
-macro toGodotName(T: typedesc): string =
-  var godotName: string
+proc toGodotName(T: typedesc): string =
```
Comment (endragor) [inline godot/nim/godotnim.nim]: Mark this explicitly as `{.compileTime.}` to make sure it's never executed in runtime.

File: godot/nim/godotnim.nim [hunk 7]
```diff
@@ -717,7 +709,7 @@ proc fromVariant*(s: var string, val: Variant): ConversionResult =
   if val.getType() == VariantType.String:
     s = val.asString()
   elif val.getType() == VariantType.Nil:
-    s = nil
+    s = ""
```
Comment (endragor) [inline godot/nim/godotnim.nim]: Should be set to `nil` in versions prior to 0.19.0.

File: godot/nim/godotmacros.nim [hunk 8]
```diff
@@ -546,15 +545,15 @@ proc genType(obj: ObjectDecl): NimNode {.compileTime.} =
                                methodNameLit, minArgs, maxArgs,
                                argTypes, methFuncIdent, hasReturnValue) =
     {.emit: """/*TYPESECTION*/
-N_NOINLINE(void, setStackBottom)(void* thestackbottom);
+N_NOINLINE(void, nimGC_setStackBottom)(void* thestackbottom);
```
Comment (endragor) [inline godot/nim/godotmacros.nim]: `setStackBottom` should still be used for versions prior to 0.19.0 here and in godotnim.nim.

**Comment by endragor** [comment]
Also, please bump the minor version.

**Comment by zetashift** [comment]
@bluenote10 any update on this, should I make a PR with the requested changes?

**Comment by bluenote10** [comment]
@zetashift That would be great, I wanted to work on it, but I simply couldn't find the time...

File: godot/nim/godotnim.nim (hunk 7 above)
Comment (zetashift) [reply godot/nim/godotnim.nim]: How would one achieve this? Using a `when` switch?

Comment (endragor) [reply godot/nim/godotnim.nim]: Yes, such as `when (NimMajor, NimMinor, NimPatch) < (0, 19, 0):`

**Comment by zetashift** [comment]
So the `{this : self}` pragma got deprecated, how should we work around this issue?

**Comment by zetashift** [comment]
Currently getting a big error and I have no idea how to solve this:
```
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
system.nim(3059)         sysFatal
Error: unhandled exception: index 1540970516 not in 0 .. -1 [IndexError]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotnim.nim:907
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
Error: unhandled exception:  [NilAccessError]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotnim.nim:907
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
[[reraised from:
godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
]]
Error: unhandled exception:  [NilAccessError]
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
system.nim(3059)         sysFatal
Error: unhandled exception: index 1540970516 not in 0 .. -1 [IndexError]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotnim.nim:907
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
Error: unhandled exception:  [NilAccessError]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotnim.nim:907
ERROR: : Unhandled Nim exception: godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
[[reraised from:
godotnim.nim(860)        godot_nativescript_init
godotnim.nim(249)        godottypes
tables.nim(860)          []=
tableimpl.nim(98)        []=
tableimpl.nim(32)        rawGet
]]
Error: unhandled exception:  [NilAccessError]
```

**Comment by jyapayne** [comment]
With the latest Nim devel as of yesterday, I can get your changes to compile and run the example code (with some modifications). I need to change `toGodotName` back to a macro and edit the `nim-godot-stub` project to remove the `isNil` checks and the mainpanel.nim to be as follows:

```nim
import godot
import scene_tree, resource_loader, packed_scene, panel, global_constants,
       input_event_mouse_button

gdobj MainPanel of Panel:
  method ready*() =
    setProcessInput(true)

  method input*(event: InputEvent) =
    if event of InputEventMouseButton:
      let ev = event as InputEventMouseButton
      if ev.buttonIndex == BUTTON_LEFT:
        getTree().setInputAsHandled()
        let scene = load("res://scene.tscn") as PackedScene
        # This line gives me a "stack smashing error" at runtime. You can also add
        # a flag to the nakefile.nim's compilation line to disable bounds checking,
        # but we should probably find out why this happens. Maybe it only happens for me?
        # getTree().root.addChild(scene.instance())
        let root = getTree().root
        let panel = root.getChild(0)
        root.addChildBelowNode(panel, scene.instance())
        queueFree()
```

I should note that I'm also using a Godot binary compiled from latest master on the godot repo.

**Comment by zetashift** [comment]
Using Godot 3.1, latest devel, compiles indeed fine(also added this on my PR to @bluenote10 branch). However the runtime error, even with your changes still happens:

```
ERROR: No loader found for resource: res://
   At: core/io/resource_loader.cpp:285
ERROR: Error loading resource: ''
   At: core/bind/core_bind.cpp:79
ERROR: : Unhandled Nim exception (NilAccessError):
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotmacros.nim:582
ERROR: No loader found for resource: res://
   At: core/io/resource_loader.cpp:285
ERROR: Error loading resource: ''
   At: core/bind/core_bind.cpp:79
ERROR: : Unhandled Nim exception (NilAccessError):
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotmacros.nim:582
ERROR: No loader found for resource: res://
   At: core/io/resource_loader.cpp:285
ERROR: Error loading resource: ''
   At: core/bind/core_bind.cpp:79
ERROR: : Unhandled Nim exception (NilAccessError):
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotmacros.nim:582
ERROR: No loader found for resource: res://
   At: core/io/resource_loader.cpp:285
ERROR: Error loading resource: ''
   At: core/bind/core_bind.cpp:79
ERROR: : Unhandled Nim exception (NilAccessError):
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]
[[reraised from:
godotmacros.nim(611)     methFunc
node.nim(54)             input
mainpanel.nim(19)        input
packed_scene.nim(74)     instance
godotnim.nim(485)        godotObject
]]

   At: ..\..\..\..\..\.nimble\pkgs\godot-0.7.19\nim\godotmacros.nim:582
```

Sorry for the lengthly log this is what the godot console gives me.

**Comment by adrianperreault** [comment]
Using Godot 3.1, latest devel, I had the same issue as @zetashift when attempting to switch scenes.

I accidentally fixed it by adding:
`  echo(repr self)
  echo(repr argToPassToGodot1)`

to the `load()` proc in `_godotapi/resource_loader.nim`

**resource_loader.nim**
```
proc load(path: string; typeHint: string = ""; noCache: bool = false): Resource =

  if isNil(singletonResourceLoader):
    singletonResourceLoader = getSingleton[ResourceLoader]()
  let self = singletonResourceLoader
  if isNil(resourceLoaderLoadMethodBind):
    resourceLoaderLoadMethodBind = getMethod(cstring"_ResourceLoader",
        cstring"load")
  var
    argsStatic: array[2, pointer]
    argsToPassToGodot = cast[ptr array[MAX_ARG_COUNT, pointer]](argsStatic.addr)
  var argToPassToGodot0 = toGodotString(path)
  argsToPassToGodot[][0] = unsafeAddr(argToPassToGodot0)
  var argToPassToGodot1 = toGodotString(typeHint)
  argsToPassToGodot[][1] = unsafeAddr(argToPassToGodot1)
  argsToPassToGodot[][2] = unsafeAddr(noCache)

  ################ Fixed Issue ???
  echo(repr self)
  echo(repr argToPassToGodot1)
  #################

  var ptrCallRet: pointer
  var ptrCallVal: ptr GodotObject
  ptrCallRet = addr(ptrCallVal)
  resourceLoaderLoadMethodBind.ptrCall(self.godotObject, argsToPassToGodot,
                                       ptrCallRet)
  deinit(argToPassToGodot0)
  deinit(argToPassToGodot1)
  result = asNimGodotObject[type(result)](ptrCallVal, false, true)

var resourceLoaderGetRecognizedExtensionsForTypeMethodBind {.threadvar.}: ptr GodotMethodBind
```

@endragor Thank you greatly for these amazing bindings! : )

File: godot/godotapigen.nim [hunk 9]
```diff
@@ -450,13 +456,13 @@ proc doGenerateMethod(tree: PNode, methodBindRegistry: var HashSet[string],
             ident("array"), ident("MAX_ARG_COUNT"),
             if isVarargs: newNode(nkPtrTy).addChain(ident("GodotVariant"))
             else: ident("pointer"))))
-      staticArgsLen = if varargsName.isNil: meth.args.len
+      staticArgsLen = if varargsName.isSome: meth.args.len
```
Comment (adrianperreault) [inline godot/godotapigen.nim]: Typo? Should it be `if not varargsName.isSome`?

**Comment by zetashift** [comment]
it's working now too, not sure why the bindings generate this :(. 
After changings the `argsStatic` I can run godot-nim on godot 3.1, 
This PR : https://github.com/godotengine/godot/pull/27320 should also make it work again without having to restart the editor

**Comment by zetashift** [comment]
@endragor do you have any idea on the off-by-one error?

**Comment by adrianperreault** [comment]
@zetashift Have you tried a clean rebuild? `nake clean` `nake build`?
With a fresh clone, latest changes the argStatic off-by-one issue went away for me.

**Comment by zetashift** [comment]
@adrianperreault yep works fine for me too now with a fresh install.
@endragor sorry for the ping, anything left to do?

**Comment by zetashift** [comment]
Currently the only annoying thing is the `{.this: } pragma is deprecated` warning. Any idea how we can fix this? I'm really not that good with macro heavy code :S.

**Comment by endragor** [comment]
Latest Nim support is added in v0.7.21.

**Pull request closed without merging**